#!/bin/bash
#$ -l h_rt=96:00:00
#$ -l mem=16G
#$ -l rmem=16G
#$ -P alloteropsis
#$ -q alloteropsis.q
#$ -pe openmp 16
#$ -v OMP_NUM_THREADS=16
#$ -j y

#####################################################################################
#       Script Name:    nuclear_assembly.sh
#       Description:    Filter out mitochondria and chloroplast reads and assemble the remaining HiFi reads
#       Author:         LPereiraG
#       Last updated:   26/10/2022
#####################################################################################

source /usr/local/extras/Genomics/.bashrc

#### Directories and files
wd=/mnt/fastdata/bo1lpg/Aristida-v3
chloroplast=${wd}/data/Aristida_chloroplast_ref.fa
reads=/shared/dunning_lab/User/bo1lpg-backup/Aristida/data/Aristida_reads.fa

#### Parameters

#### Step 1: Generate a FASTA with the mitochondrial selected contigs from multi GFF3+FASTA file
# The GFF3+FASTA is generated by GeSeq and contains only contigs with annotated genes.
# Use those contig names to extract the sequences and generate a final mitochondrial draft genome
# cd A02_results
#mkdir final_contigs_after_annotation
# cd final_contigs_after_annotation
# grep ">" Aristida_contigs_GLOBAL_multi-GFF3+FASTA.gff3 | sed 's/>//g' > contig.list
# seqtk subseq ${wd}/A02_results/reads_mapping_and_assembly/hifiasm.contigs.clean.fasta contig.list \
#   >> ${wd}/data/Aristida_mitochondria_ref.fa
# mitochondria=${wd}/data/Aristida_mitochondria_ref.fa

#### Step 2: Concatenate the two organelle references and map all reads against it. Select unmapped reads.
cd ${wd}
# mkdir A03_results
cd A03_results
# module load apps/python/anaconda2-4.2.03
# source activate /shared/dunning_lab/Shared/conda_env/minimap2
# cat ${mitochondria} ${chloroplast} >> organelle_ref.fa
# minimap2 -x map-hifi organelle_ref.fa ${reads} -t 4 > aln_organelle.paf
# awk '{if ($10/$2 > 0.8) print $1}' aln_organelle.paf | sort | uniq > reads_to_filter_out.txt
# cp ${reads} Aristida_reads.fasta
# samtools faidx Aristida_reads.fasta
# awk '{print $1}' Aristida_reads.fasta.fai >> all_reads.txt
# cat all_reads.txt reads_to_filter_out.txt | sort | uniq -u >> reads_to_keep.txt
seqtk subseq Aristida_reads.fasta reads_to_keep.txt >> Aristida_nuclear_reads.fa

#### Step 3: Assemble the nuclear reads
/shared/dunning_lab/Shared/programs/hifiasm-0.16.1/./hifiasm -o Aristida_nuclear.asm -t 16 Aristida_nuclear_reads.fa
